shared/src/balancer/swap/fixed_point.rs

```diff
@@ -1,3 +1,307 @@
+    }
+
+    pub fn mul_down(self, other: Self) -> Result<Self, Error> {
+        Ok(Self(
+            self.0.checked_mul(other.0).ok_or(Error::MulOverflow)? / *ONE_18,
+        ))
+    }
+
+    pub fn mul_up(self, other: Self) -> Result<Self, Error> {
+        let product = self.0.checked_mul(other.0).ok_or(Error::MulOverflow)?;
+
+        Ok(if product.is_zero() {
+            Bfp::zero()
+        } else {
+            Bfp(((product - 1) / *ONE_18) + 1)
+        })
+    }
+
+    pub fn div_down(self, other: Self) -> Result<Self, Error> {
+        if other.is_zero() {
+            Err(Error::ZeroDivision)
+        } else {
+            Ok(Self(
+                self.0.checked_mul(*ONE_18).ok_or(Error::DivInternal)? / other.0,
+            ))
+        }
+    }
+
+    pub fn div_up(self, other: Self) -> Result<Self, Error> {
+        if other.is_zero() {
+            return Err(Error::ZeroDivision);
+        }
+        if self.is_zero() {
+            Ok(Self::zero())
+        } else {
+            let a_inflated = self.0.checked_mul(*ONE_18).ok_or(Error::DivInternal)?;
+
+            Ok(Self(((a_inflated - 1) / other.0) + 1))
+        }
+    }
+
+    pub fn complement(self) -> Self {
+        if self.0 < *ONE_18 {
+            Self(*ONE_18 - self.0)
+        } else {
+            Self::zero()
+        }
+    }
+
+    pub fn pow_up(self, exp: Self) -> Result<Self, Error> {
+        let raw = Bfp(logexpmath::pow(self.0, exp.0)?);
+        let max_error = raw.mul_up(*MAX_POW_RELATIVE_ERROR)?.add(Bfp(1.into()))?;
+
+        raw.add(max_error)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn parsing() {
+        assert_eq!("1".parse::<Bfp>().unwrap(), Bfp::one());
+        assert_eq!(
+            "0.1".parse::<Bfp>().unwrap(),
+            Bfp::from_wei(U256::exp10(17))
+        );
+        assert_eq!(
+            "1.01".parse::<Bfp>().unwrap(),
+            Bfp::from_wei(U256::exp10(18) + U256::exp10(16))
+        );
+        assert_eq!(
+            "10.000000000000000001".parse::<Bfp>().unwrap(),
+            Bfp::from_wei(U256::exp10(19) + U256::one())
+        );
+        assert!("10.0000000000000000001".parse::<Bfp>().is_err());
+        assert!("1.0.1".parse::<Bfp>().is_err());
+        assert!(".1".parse::<Bfp>().is_err());
+        assert!("1.".parse::<Bfp>().is_err());
+        assert!("".parse::<Bfp>().is_err());
+    }
+
+    #[test]
+    fn add() {
+        assert_eq!(Bfp::from(40).add(2.into()).unwrap(), 42.into());
+
+        assert_eq!(
+            Bfp::MAX.add(Bfp::epsilon()).unwrap_err(),
+            Error::AddOverflow
+        );
+    }
+
+    #[test]
+    fn sub() {
+        assert_eq!(Bfp::from(50).sub(8.into()).unwrap(), 42.into());
+
+        assert_eq!(
+            Bfp::one().sub(Bfp(*ONE_18 + 1)).unwrap_err(),
+            Error::SubOverflow
+        );
+    }
+
+    macro_rules! test_mul {
+        ($fn_name:ident) => {
```

Comment:
Maybe a test where mul|div up|down return different values?

Reply:
I tested the differerent behavior [here](https://github.com/gnosis/gp-v2-services/pull/740/files#diff-becc6a22d1880dbaa103dad137a9f7b67cb8a7f5276329373edd20f459d1b0ddR231-R233) for `mul` and [here](
https://github.com/gnosis/gp-v2-services/pull/740/files#diff-becc6a22d1880dbaa103dad137a9f7b67cb8a7f5276329373edd20f459d1b0ddR259-R260) for `div`. I used `.epsilon()` because it was easier to come up with a test case, but maybe I should use a proper number?